File: v4_law_of_demeter/java/SurfShop2.java
public class SurfShop2 {

    public static void main(String[] args) {
        SurfShop2 surfShop2 = new SurfShop2();
        Customer customer = new Customer();
        surfShop2.chargeCustomer(customer, 10);
        System.out.println("Done");
    }

    public void chargeCustomer(Customer c, float fee) {
        c.pay(fee);
    }
}

class Customer {

    private Card card = new Card();

    public void pay(float fee) {
        this.card.deduct(fee);
    }

}

class Card {

    public int balance = 10;

    public void deduct(float fee) {
        this.balance -= fee;
    }

}

File: v4_law_of_demeter/python/surfshop2.py
class Card:

    balance = 10

    def deduct(self, fee: float): 
        self.balance -= fee

class Customer:

    card = Card()

    def pay(self, fee: float):
        self.card.deduct(fee)

class SurfShop2:

    def charge_customer(self, c: Customer, fee: float):
        c.pay(fee)

customer = Customer()
SurfShop2().charge_customer(customer, 1.0)

